rfcs/002-references-and-pointers.md

```diff
@@ -0,0 +1,332 @@
+```
+
+Below is a pointer, it functions equivalently with mutability however assignments are different. This is the current behavior in V.
+
+```vlang
+struct Struct {
+	/* ... */
+}
+
+fn take_struct(mut v &Struct) {
+	ptr := &Struct(/* ... */)
+
+	// pointer arithmetic, may be nil in unsafe contexts.
+	// it's a pointer, overwrite memory using *
+	*v = Struct{/* ... */}
+
+	field := v.field // auto deref on fields, this is normal
+	v = ptr          // assigning integers to integers..
+}
+```
+
+With the above, it's understood that the second example now explicitly takes a pointer, and the first example is just a variable that can be edited.
+
+---
+
+Instead of thinking of `mut params` as pointers, we can be explicit here.
+
+The parameter of the `fn send` is mutable, it's a mutable binding to a variable outside of the function. We can take the address of it with the `&` operator, and it's the address of that outside variable.
+
+```vlang
+struct AA {/* ... */}
+
+// the `mut AA` doesn't make sense for external C functions.
+// fn C.external(mut AA)
+
+fn C.external(mut &AA)
+
+fn send(mut value AA) {
+	C.external(&value)
+}
+```
+
+You can't cast a `mut param` to a pointer, because it isn't a pointer. You must take the address of it.
+
+No need to explicitly construct a pointer value, if you allow V to promote it's allocation you won't leave auto-deref territory.
+
+You shouldn't even think in "promotion of allocations." At the end of the day a variable is a variable, and pointers are hidden.
+
+```vlang
+struct AA {/* ... */}
+
+fn return_aa(value AA) &AA {
+	mut a := AA{}
+	a.field = 10
+	a = value     // "auto deref" -- this isn't a pointer to users
+	return a
+}
+```
+
+**What about immutable `&T` ?**
+
+When a `T` is immutable, it doesn't matter if it is passed by value or by reference. There is zero difference in underlying assembly code. Immutability plays a big part here.
+
+**90% of the time, you will never use a `&T` in V code. It should be discouraged.**
+
+If you're working with linked lists, need to set pointer values, that's fine. However, if you're working with an immutable receiver it doesn't matter.
+
+Specifying a `&T` or `mut &T` will turn off auto-deref, for 90% of V usage, we should keep this. Keeping auto-deref means a `&T` shouldn't be used.
+
+```vlang
+fn (ref T) test0() {
+	// ref is just a variable
+	// it's immutable anyway, we don't need &T
+}
+
+fn (ref &T) test1() {
+	// ref is a pointer, do arithmetic, etc
+}
+```
+
+**You still want to work on pointers with auto-deref?**
+
+Auto-deref is the current behavior with pointers, to keep operations on them the same as normal variables. Using the `[heap]` attribute or allowing the local variable to escape will promote them to become heap allocated, therefore you can still work on them just like normal variables.
+
+```vlang
+[heap]
+struct AA {}
+
+fn test() {
+	// a is a variable, pointers are abstracted away.
+	mut a := AA{}
+}
+
+struct BB {}
+
+fn test_ptr() &BB {
```

Comment:
Why return a pointer here?

Reply:
To allow V to promote the allocation of the `BB` struct to the heap, with all accesses in V acting like there are no pointers involved. It's hiding/abstracting away pointers.

Reply:
But I think the you should have to return like `&bb`